File: readers_writers.py
import threading
import time

READERS_COUNT = 5
WRITERS_COUNT = 3

lock = threading.Lock()
cv_read = threading.Condition(lock)
cv_write = threading.Condition(lock)
read_count = 0
write_count = 0
data = 0

def reader(id):
    global read_count, data
    
    while True:
        # Enter reader critical section
        with lock:
            while write_count > 0:
                cv_read.wait()
            read_count += 1
        
        # Read data
        print(f"Reader {id} reads data {data}")
        time.sleep(0.5)
        
        # Exit reader critical section
        with lock:
            read_count -= 1
            if read_count == 0:
                cv_write.notify()
        
        # Sleep for some time
        time.sleep(0.5)

def writer(id):
    global write_count, data
    
    while True:
        # Enter writer critical section
        with lock:
            while write_count > 0 or read_count > 0:
                cv_write.wait()
            write_count += 1
        
        # Write data
        data += 1
        print(f"Writer {id} writes data {data}")
        time.sleep(0.5)
        
        # Exit writer critical section
        with lock:
            write_count -= 1
            if write_count == 0:
                cv_read.notify_all()
                cv_write.notify()
        
        # Sleep for some time
        time.sleep(0.5)

# Start reader threads
readers = []
for i in range(READERS_COUNT):
    readers.append(threading.Thread(target=reader, args=(i,)))
    readers[-1].start()

# Start writer threads
writers = []
for i in range(WRITERS_COUNT):
    writers.append(threading.Thread(target=writer, args=(i,)))
    writers[-1].start()

# Wait for all threads to finish
for r in readers:
    r.join()
for w in writers:
    w.join()

File: readers_writers.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
using namespace std;

const int READERS_COUNT = 5;
const int WRITERS_COUNT = 3;

mutex mtx;
condition_variable cv_read, cv_write;
int read_count = 0, write_count = 0, data = 0;

void reader(int id) {
    while (true) {
        unique_lock<mutex> lock(mtx);
        while (write_count > 0) {
            cv_read.wait(lock);
        }
        ++read_count;
        lock.unlock();
        
        cout << "Reader " << id << " reads data " << data << endl;
        this_thread::sleep_for(chrono::milliseconds(500));
        
        lock.lock();
        --read_count;
        if (read_count == 0) {
            cv_write.notify_one();
        }
        lock.unlock();
        
        this_thread::sleep_for(chrono::milliseconds(500));
    }
}

void writer(int id) {
    while (true) {
        
        unique_lock<mutex> lock(mtx);
        while (write_count > 0 || read_count > 0) {
            cv_write.wait(lock);
        }
        ++write_count;
        lock.unlock();
        
        ++data;
        cout << "Writer " << id << " writes data " << data << endl;
        this_thread::sleep_for(chrono::milliseconds(500));
        
        lock.lock();
        --write_count;
        if (write_count == 0) {
            cv_read.notify_all();
            cv_write.notify_one();
        }
        lock.unlock();
        
        this_thread::sleep_for(chrono::milliseconds(500));
    }
}

int main() {
    thread readers[READERS_COUNT];
    thread writers[WRITERS_COUNT];
    
    for (int i = 0; i < READERS_COUNT; ++i) {
        readers[i] = thread(reader, i);
    }
    
    for (int i = 0; i < WRITERS_COUNT; ++i) {
        writers[i] = thread(writer, i);
    }
    
    for (int i = 0; i < READERS_COUNT; ++i) {
        readers[i].join();
    }
    for (int i = 0; i < WRITERS_COUNT; ++i) {
        writers[i].join();
    }
    
    return 0;
}
